Refine MetricProvider.ForceFlush and define ForceFlush for periodic exporting MetricReader

pellared opened the pull request:
It looks like we have forgotten to define `ForceFlush` for metric readers. Take notice that the spec currently references it here: 

https://github.com/open-telemetry/opentelemetry-specification/blob/c07f4a2616ff50c7189f3365be20ef090ad53fc4/specification/metrics/sdk.md#L173-L175

This PR is intended to make the specification around metric's `ForceFlush` clearer and more normative. 

The proposal does not require each `MetricReader` to implement `ForceFlush` (because it not needed for pull-based exporters) This is intended to support the following:
- https://github.com/open-telemetry/opentelemetry-go/pull/4375
- https://github.com/open-telemetry/opentelemetry-php/issues/1086#issuecomment-1662140452

reyang commented on specification/metrics/sdk.md:
```diff
@@ -1373,7 +1373,7 @@ called concurrently.
 **ExemplarReservoir** - all methods are safe to be called concurrently.
 
-**MetricReader** - `Collect` and `Shutdown` are safe to be called concurrently.
+**MetricReader** - `Collect`, `ForceFlush` and `Shutdown` are safe to be called concurrently.
```
Where is this? https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader-operations

![image](https://github.com/open-telemetry/opentelemetry-specification/assets/17327289/2875b937-957e-476c-ac0a-3b1094b1f25b)

pellared replied:
This fragment caused confusion:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md?plain=1#L171-L173

But you are right that the SDK does not explicitly require to have this method.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md?plain=1#L979-L983

But what is very strange is that ☝️  says that it accepts `OnShutdown callback functions.` and then we have https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md?plain=1#L1023-L1037

Do you have any proposal how to fix it?

My initial proposal would be to add a `ForceFlush` to the `MetricReader operations` section as I believe that deciding whether the operation is handled via callback or direct call is an implementation detail (but I may be wrong).

reyang replied:
Similar to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shutdown-1

_Shutdown MUST include the effects of ForceFlush._

I think we can change _"ForceFlush MUST invoke ForceFlush on all registered MetricReader and Push Metric Exporter instances"_ to _"ForceFlush MUST include the effects of Collect on all registered MetricReader instances, and ForceFlush on all Push Metric Exporter instances"_.

jmacd replied:
This conversation looks independent of the change in this PR, is that right?

pellared replied:
I think it makes sense to address @reyang comment to make the PR more complete. These are "cohesive" changes.

pellared replied:
@reyang Thanks for your suggestions. PTAL

reyang reviewed (changes requested):
https://github.com/open-telemetry/opentelemetry-specification/pull/3563/files#r1239953953

reyang commented on specification/metrics/sdk.md:
```diff
@@ -1038,6 +1039,21 @@ implemented as a blocking API or an asynchronous API which notifies the caller
 via a callback or an event. [OpenTelemetry SDK](../overview.md#sdk) authors MAY
 decide if they want to make the shutdown timeout configurable.
 
+#### ForceFlush
```
@pellared do you feel we have a clear understanding regarding "what's the difference between MetricReader.Collect and MetricReader.ForceFlush, and why do we need MetricReader.ForceFlush"? My gut feeling is that we don't.

pellared replied:
I think this "kind of" duplicates. https://github.com/open-telemetry/opentelemetry-specification/pull/3563#discussion_r1245592766. Please resolve if you agree 😉 Otherwise I need more clarification.

reyang replied:
I feel we're heading towards the wrong direction:
1. The change proposed in this PR doesn't solve the problem in https://github.com/open-telemetry/opentelemetry-specification/blob/c07f4a2616ff50c7189f3365be20ef090ad53fc4/specification/metrics/sdk.md#L173-L175
2. It's unclear what's the responsibility of `Collect` and `ForceFlush` for periodic exporting MetricReader, and what exactly is expected from `MeterProvider.ForceFlush`.

pellared replied:
I agree with you.

I think that we should say that basically periodic reader's force flush does something like this:
Pseudocode:

```
metrics = this.Collect() 
this.exporter.Export(metrics)
this.exporter.ForceFlush()
```

Do you agree?

pellared replied:
Now I think I understand this https://github.com/open-telemetry/opentelemetry-specification/pull/3563#discussion_r1242732786 😅

reyang replied:
I disagree 😆

Export will yield the Success/Failure result, which means the subsequent call to exporter.ForceFlush would make no sense.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exportbatch

pellared replied:
I think it may be better if you address the issue in a separate PR. I think the comment is not conflicting with existing PR.

pellared replied:
Maybe https://github.com/open-telemetry/opentelemetry-specification/pull/3609 😉

MrAlias commented on specification/metrics/sdk.md:
```diff
@@ -1038,6 +1039,21 @@ implemented as a blocking API or an asynchronous API which notifies the caller
 via a callback or an event. [OpenTelemetry SDK](../overview.md#sdk) authors MAY
 decide if they want to make the shutdown timeout configurable.
 
+#### ForceFlush
```
It doesn't make sense to define this for all `MetricReader`, it should only apply to the periodic exporting `MetricReader`. A pull based setup should not be required to implement this method as it will be a no-op and a user should call `Collect` instead.

pellared replied:
I moved it under `Periodic exporting MetricReader` section

https://github.com/open-telemetry/opentelemetry-specification/pull/3563/commits/2b8f91e5332c93a7f1c48c855b4b7266803b0abb

pellared commented:
I have currently no bandwidth to work on it 😢 Also I am not sure if I understand @reyang suggestion. 

@reyang are you able to propose a PR yourself or make changes in this PR?

pellared commented:
@carlosalberto I created https://github.com/open-telemetry/opentelemetry-specification/pull/3609

@reyang I think this is what you had in mind 😉 PTAL

pellared commented:
Closing. See https://github.com/open-telemetry/opentelemetry-specification/pull/3563

pellared commented on specification/metrics/sdk.md:
```diff
@@ -1183,6 +1183,25 @@ from `MetricReader` and start a background task which calls the inherited
+This method provides a way for the periodic exporting MetricReader
+so it can do as much as it could to collect and send the metrics.
+
+`ForceFlush` SHOULD collect metrics, call [`Export(batch)`](#exportbatch)
```
My preference is

```suggestion
`ForceFlush` MUST collect metrics, call [`Export(batch)`](#exportbatch)
```

Aneurysm9 replied:
I believe this should be left as `SHOULD` to allow for no-op and pass-through implementations.

pellared commented on specification/metrics/sdk.md:
```diff
@@ -160,13 +161,16 @@ decide if they want to make the shutdown timeout configurable.
 data when it is being asked by the scraper, so `ForceFlush` would not make much
 sense.
 
+`ForceFlush` MUST invoke `ForceFlush` on all registered
+[Periodic exporting MetricReader](#periodic-exporting-metricreader) instances.
```
Alternative version

```suggestion
`ForceFlush` MUST invoke `ForceFlush` on all registered
[MetricReader](#metricreader) instances that implement `ForceFlush`.
```

This would be more generic if the SDK provides more implementation of readers that handle `ForceFlush`. It may also be easier to interpret for SDK that prefer to add `ForceFlush` to the reader interface and implement it as a no-op e.g. for other reader implementations.

Aneurysm9 replied:
I think that all `MetricReader` implementations should be required to implement `ForceFlush`, even if only as a no-op, and the `MeterProvider` must invoke that method on all registered readers.  This is the same thing we do for [`SpanProcessor` implementations](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush-1).

pellared replied:
This is IMO language specific if metric readers have to implement it as a NOOP or just not implement it.  From PR description:

> The proposal does not require each MetricReader to implement ForceFlush (because it not needed for pull-based exporters) This is intended to support the following:
> 
> https://github.com/open-telemetry/opentelemetry-go/pull/4375
> https://github.com/open-telemetry/opentelemetry-php/issues/1086#issuecomment-1662140452

pellared commented:
Reopened per https://github.com/open-telemetry/opentelemetry-specification/pull/3609#issuecomment-1661755743.

I updated the PR, based on the feedback I got so far.

I added a comment here: https://github.com/open-telemetry/opentelemetry-specification/issues/1432#issuecomment-1661701778. I do not think that this PR conflicts with https://github.com/open-telemetry/opentelemetry-specification/issues/1432.

pellared commented:
@jsuereth @reyang PTAL

brettmc commented on specification/metrics/sdk.md:
```diff
@@ -1183,6 +1183,25 @@ from `MetricReader` and start a background task which calls the inherited
 
+#### ForceFlush
+
+This method provides a way for the periodic exporting MetricReader
+so it can do as much as it could to collect and send the metrics.
```
This sounds clunky, how about this? 
```suggestion
This is a hint to a  periodic exporting MetricReader that it SHOULD do as much as possible
to collect and send metrics as soon as possible, preferably before returning from this method.
```

pellared replied:
I would rather prefer to do it in a separate PR and also change similar sections like:

https://github.com/open-telemetry/opentelemetry-specification/blob/b2f923fb1650dde1f061507908b834035506a796/specification/metrics/sdk.md#L162-L164

Is it a blocker for you?

brettmc replied:
ok, it's not a blocker for me.
